### PR TITLE
Update lofreq somatic and callparallel 

### LINKED ISSUE #5930: Update lofreq somatic & lofreq call-parallel
### Have you checked the docs?

- [X] [nf-core website: troubleshooting](https://nf-co.re/usage/troubleshooting)
- [X] [nf-core modules documentation](https://nf-co.re/docs/contributing/modules)

### Description of the bug

Lofreq somatic & lofreq call parallel modules only work if they recieve a `bam` file as an input. @AitorPeseta and I will be working in updating these modules to convert `cram` files to `bam` bam format, since we are planning to contribute to Sarek by adding this caller.

### Command used and terminal output

_No response_

### Relevant files

_No response_

### System information

_No response_

### PR DESCRIPTION
**Pull request by AitorPeseta**
## PR checklist

Closes #5930 

Lofreq somatic & lofreq call parallel modules only work if they recieve a bam file as an input.  We will be working in updating these modules to convert cram files to bam bam format, since we are planning to contribute to Sarek by adding this caller. We have taken the code from cnvkit batch module.

- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - For modules:
    - [x] `nf-core modules test <MODULE> --profile docker`
    - [x] `nf-core modules test <MODULE> --profile singularity`
    - [x] `nf-core modules test <MODULE> --profile conda`

### CODE REVIEW

**Comment by nevinwu** [comment]
@nf-core-bot fix linting pretty please

**Comment by famosab** [comment]
While you are working on that: Could you also port the modules to nf-test? And add stubs? 🚀

File: modules/nf-core/lofreq/callparallel/main.nf [hunk 1]
```diff
@@ -8,7 +8,7 @@ process LOFREQ_CALLPARALLEL {
         'biocontainers/lofreq:2.1.5--py38h588ecb2_4' }"
 
     input:
-    tuple val(meta), path(bam), path(bai), path(intervals)
+    tuple val(meta), path(tumor), path(tumor_index), path(intervals)
```
Comment (famosab) [inline modules/nf-core/lofreq/callparallel/main.nf:11]: Why did you change the inputs?

File: modules/nf-core/lofreq/callparallel/meta.yml [hunk 2]
```diff
@@ -18,11 +18,11 @@ input:
       description: |
         Groovy Map containing sample information
         e.g. [ id:'test' ]
-  - bam:
+  - tumor:
       type: file
       description: Sorted BAM file
```
Comment (famosab) [inline modules/nf-core/lofreq/callparallel/meta.yml:23]:
```suggestion
      description: Tumor sample sorted BAM file
```

File: modules/nf-core/lofreq/somatic/meta.yml [hunk 3]
```diff
@@ -23,19 +23,19 @@ input:
...
-  - tumor_bam:
+  - tumor:
       type: file
       description: tumor sample input BAM file
       pattern: "*.{bam}"
-  - tumor_bai:
+  - tumor_index:
       type: file
       description: tumor sample BAM index file
       pattern: "*.{bai}"
```
Comment (famosab) [inline modules/nf-core/lofreq/somatic/meta.yml:33]:
```suggestion
      pattern: "*.{bam.bai}"
```

File: modules/nf-core/lofreq/somatic/meta.yml [hunk 4]
```diff
@@ -23,19 +23,19 @@ input:
...
-  - normal_bam:
+  - normal:
       type: file
       description: normal sample input BAM file
       pattern: "*.{bam}"
-  - normal_bai:
+  - normal_index:
       type: file
       description: normal sample BAM index file
       pattern: "*.{bai}"
```
Comment (famosab) [inline modules/nf-core/lofreq/somatic/meta.yml:41]:
```suggestion
      pattern: "*.{bam.bai}"
```

File: modules/nf-core/lofreq/callparallel/meta.yml [hunk 5]
```diff
@@ -18,11 +18,11 @@ input:
...
       pattern: "*.{bam}"
-  - bai:
+  - tumor_index:
       type: file
       description: BAM index file
       pattern: "*.{bai}"
```
Comment (famosab) [inline modules/nf-core/lofreq/callparallel/meta.yml:28]:
```suggestion
      description: Tumor sample sorted BAM index file
      pattern: "*.{bam.bai}"
```

File: modules/nf-core/lofreq/callparallel/main.nf [hunk 6]
```diff
@@ -23,15 +23,25 @@ process LOFREQ_CALLPARALLEL {
...
+
+    def tumor_cram =  tumor.Extension == "cram" ? true : false
+    def tumor_bam = tumor.Extension == "bam" ? true : false
+    def tumor_out = tumor_cram ? tumor.BaseName + ".bam" : "${tumor}"
+
+    def samtools_cram_convert = ''
+    samtools_cram_convert += tumor_cram ? "    samtools view -T $fasta $tumor -@ $task.cpus -o $tumor_out\n" : ''
+    samtools_cram_convert += tumor_cram ? "    samtools index $tumor_out\n" : ''
     """
+    $samtools_cram_convert
```
Comment (SPPearce) [inline modules/nf-core/lofreq/callparallel/main.nf:35]: Could this be piped into lofreq, rather than having to make a copy?

File: modules/nf-core/lofreq/callparallel/main.nf [hunk 7]
```diff
@@ -23,15 +23,25 @@ process LOFREQ_CALLPARALLEL {
     def args = task.ext.args ?: ''
     def prefix = task.ext.prefix ?: "${meta.id}"
     def options_intervals = intervals ? "-l ${intervals}" : ""
+
+    def tumor_cram =  tumor.Extension == "cram" ? true : false
+    def tumor_bam = tumor.Extension == "bam" ? true : false
+    def tumor_out = tumor_cram ? tumor.BaseName + ".bam" : "${tumor}"
+
+    def samtools_cram_convert = ''
```
Comment (SPPearce) [inline modules/nf-core/lofreq/callparallel/main.nf:31]: If you are converting the files, please remove them at the end, don't want 50Gb bam files lying around...

**Comment by SPPearce** [comment]
Please make sure you clean up the files, don't want to leave them taking up space afterwards.
Why are you renaming the inputs in `lofreq callparallel`? I imagine this could be used for single sample variant calling, which is not necessarily tumour.

**Comment by AitorPeseta** [comment]
Hi, we have already implemented the suggested changes, including updating meta.yml (description and pattern) and cleaning up the files. However, there is one thing we couldn't do: the nf-test. We are not familiar with nf-test and don't have time. I apologize for this.

File: modules/nf-core/lofreq/callparallel/main.nf [hunk 8]
```diff
@@ -23,15 +23,42 @@ process LOFREQ_CALLPARALLEL {
...
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        lofreq: \$(echo \$(lofreq version 2>&1) | sed 's/^version: //; s/ *commit.*\$//')
+    END_VERSIONS
+    """
+
+    stub:
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    """
+    touch ${prefix}.vcf.gz
```
Comment (famosab) [inline modules/nf-core/lofreq/callparallel/main.nf:61]:
```suggestion
    echo "" | gzip > ${prefix}.vcf.gz
```

**Review by famosab — COMMENTED** [review]
@AitorPeseta I would suggest you look into it, it has to be done at some point. I can help with completing. Its not a lot of work and nf-core tools will help you do the porting. What you do is basically
```
nf-core modules create lofreq/somatic --migrate-pytest
nf-core modules create lofreq/callparallel --migrate-pytest
```
This will create all files for nf-test and remove from pytest. Then you will just need to adjust the `main.nf.test` files by copying most of the input from the pytest files. And afterwards run 
```
nf-core modules test lofreq/somatic --profile <docker or conda> -u
nf-core modules test lofreq/callparallel --profile <docker or conda> -u
```
to create the snapshot. But I understand if there is no time to do it now!

**Comment by famosab** [comment]
There is work on `lofreq/indelqual` port to nf-test: #5941

**Comment by SPPearce** [comment]
> @AitorPeseta I would suggest you look into it, it has to be done at some point. I can help with completing. Its not a lot of work and nf-core tools will help you do the porting. What you do is basically
> 
> ```
> nf-core modules create lofreq/somatic --migrate-pytest
> nf-core modules create lofreq/callparallel --migrate-pytest
> ```
> 
> This will create all files for nf-test and remove from pytest. Then you will just need to adjust the `main.nf.test` files by copying most of the input from the pytest files. And afterwards run
> 
> ```
> nf-core modules test lofreq/somatic --profile <docker or conda> -u
> nf-core modules test lofreq/callparallel --profile <docker or conda> -u
> ```
> 
> to create the snapshot. But I understand if there is no time to do it now!

I have started the process on somatic (not pushed yet). It is slightly tricky here, as all the vcf files contain the directory that was used to generate them (and the date), so the assertions are harder.

**Comment by famosab** [comment]
> I have started the process on somatic (not pushed yet). It is slightly tricky here, as all the vcf files contain the directory that was used to generate them (and the date), so the assertions are harder.

Ok then I think this PR is good to go without the porting!

**Comment by AitorPeseta** [comment]
Hi @SPPearce , will your push be in this request or in another one? Also, our request is Auto-merge enabled, so we think it needs your approval. Thanks for your help.

**Comment by SPPearce** [comment]
I'll push to here once I have the test working locally

**Comment by famosab** [comment]
I can try and port callparallel 👍

**Comment by SPPearce** [comment]
Can we add the meta map to the fasta and fai inputs for callparallel while we are here

**Review by SPPearce — COMMENTED** [review]
In theory we should swap the test file paths in somatic, to use the new format.

**Comment by famosab** [comment]
@SPPearce done 🚀

File: modules/nf-core/lofreq/somatic/tests/main.nf.test [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+                    process.out.vcf[0][1].size(),
+                    process.out.versions
+                    ).match()
+                }
+            )
+        }
+
+    }
+
+    test("sarscov2 - bam - stub") {
```
Comment (SPPearce) [inline modules/nf-core/lofreq/somatic/tests/main.nf.test:51]:
```suggestion
    test("human - bam - stub") {
```

**Review by SPPearce — APPROVED** [review]

**Comment by nevinwu** [comment]
Thank you @famosab @SPPearce from Aitor and me for your feedback and work on the nf-test part!
